### ISSUE TITLE
download speed drops to 0 using streamProvider

### ISSUE THREAD

**Issue by borrelnoten** [alanmcgovern/monotorrent#373]
version 31

many torrent stuck while downloading with 0 bytes per seconds download at around 30% or 50% ot 60% or sometimes 24% using streamProvider. This is some time after seeking to surtain byte position in the file I need to download.

**Comment by borrelnoten** [comment]
after 5 minutes waiting, speeds seems to be picking up. Seems the speed either spikes or drops very much after some time. Does that makes sense?

**Comment by alanmcgovern** [comment]
It's interesting that there are issues with 31.  The changes I made with rev37 have changed how pieces are selected (again) and so the results with rev31 are no longer good enough to predict the behaviour of rev37+.

I've tried a few torrents and so far so good with rev37. I'll keep trying with rev37+ for now!

**Comment by borrelnoten** [comment]
I made a mistake, I am using version 37.

**Comment by borrelnoten** [comment]
It seems getting the last piece of a file is difficult

**Comment by alanmcgovern** [comment]
I've landed a bucket of changes to piece picking. I believe you were hitting some issues in the original changes I had made when I added the concept of 'abandoned' pieces to the piece picker. The bug allowed the same blocks to be requested multiple times if a piece were marked as 'abandoned', as opposed to allowing any peer to request each block once.

The new features should be covered in unit tests now, and this scenario shouldn't occur!
